docs/reference/python/conventions.md

```diff
@@ -36,14 +36,98 @@ and future members of the software team. The major things that we document in ou
 Ideally, the third point should be avoided as much as possible since we would want our code to be
 self explanatory. It should be done only when absolutely necessary.
 
+### Type hinting and static type checking
+
+Since Python is a dynamically typed language, we need to make use of the [`typing`](https://docs.python.org/3/library/typing.html){target=_blank}
+and [`mypy`](https://mypy.readthedocs.io/en/stable/index.html){target=_blank} modules if we want to do
+any type checking in our functions and classes. Type checking is beneficial for documentation and maintaining a clean
+software architecture. There is some syntax to get familiar in order to use type checking. We recommend the following
+resources:
+
+- [mypy Typing Cheatsheet](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html){target=_blank}
+- [PEP 483: The Theory of Type Hints (A Simplified Guide)](https://peps.python.org/pep-0483/){target=_blank}
+- [PEP 484: Type Hints (Fully Comprehensive Guide)](https://peps.python.org/pep-0484/){target=_blank}
+
+You can check for type mismatches in the command line by running `mypy myfile.py`. Below are a few examples of
+using type hinting:
+
+??? example "Return the sum of a sequence"
+
+    ```python
+    from typing import Sequence, Union
+    Number = Union[int, float]
+
+    def sumseq(seq : Sequence[Number]) -> Number:
+        return sum(seq)
```

Comment:
Code examples should pass flake8
- 2 spaces between the imports and constants sections, as well as between any classes or global functions